File: front_middle_back_queue.dart
class FrontMiddleBackQueue {
  List<int> nums = [];

  FrontMiddleBackQueue() {}

  void pushFront(int val) {
    nums.insert(0, val);
  }

  void pushMiddle(int val) {
    int length = nums.length;
    nums.insert(length ~/ 2, val);
  }

  void pushBack(int val) {
    nums.add(val);
  }

  int popFront() {
    if (nums.isEmpty) {
      return -1;
    }
    return nums.removeAt(0);
  }

  int popMiddle() {
    if (nums.isEmpty) {
      return -1;
    }
    int length = nums.length;
    if (length % 2 == 0) {
      return nums.removeAt(length ~/ 2 - 1);
    } else {
      return nums.removeAt(length ~/ 2);
    }
  }

  int popBack() {
    if (nums.isEmpty) {
      return -1;
    }
    return nums.removeLast();
  }
}

/**
 * Your FrontMiddleBackQueue object will be instantiated and called as such:
 * FrontMiddleBackQueue obj = FrontMiddleBackQueue();
 * obj.pushFront(val);
 * obj.pushMiddle(val);
 * obj.pushBack(val);
 * int param4 = obj.popFront();
 * int param5 = obj.popMiddle();
 * int param6 = obj.popBack();
 */

File: front_middle_back_queue.py
import collections


class FrontMiddleBackQueue:

    def __init__(self):
        self.left = collections.deque()
        self.right = collections.deque()

    def pushFront(self, val: int) -> None:
        self.left.appendleft(val)
        if len(self.left) == len(self.right) + 2:
            self.right.appendleft(self.left.pop())

    def pushMiddle(self, val: int) -> None:
        if len(self.left) == len(self.right) + 1:
            self.right.appendleft(self.left.pop())
        self.left.append(val)

    def pushBack(self, val: int) -> None:
        self.right.append(val)
        if len(self.left) + 1 == len(self.right):
            self.left.append(self.right.popleft())

    def popFront(self) -> int:
        if len(self.left) == 0:
            return -1
        val = self.left.popleft()
        if len(self.left) + 1 == len(self.right):
            self.left.append(self.right.popleft())
        return val

    def popMiddle(self) -> int:
        if len(self.left) == 0:
            return -1
        val = self.left.pop()
        if len(self.left) + 1 == len(self.right):
            self.left.append(self.right.popleft())
        return val

    def popBack(self) -> int:
        if len(self.left) == 0:
            return -1
        val = 0
        if len(self.right) == 0:
            val = self.left.pop()
        else:
            val = self.right.pop()
            if len(self.left) == len(self.right) + 2:
                self.right.appendleft(self.left.pop())
        return val